akka-cluster-typed/src/test/scala/akka/cluster/typed/GroupRouterSpec.scala

```diff
@@ -31,154 +35,165 @@ object GroupRouterSpec {
 
-  val pingPongKey = ServiceKey[Ping.type]("ping-pong")
+  val pingPongKey = ServiceKey[PingActor.Ping]("ping-pong")
+
+  case class Result(actorsInSystem1: Int, actorsInSystem2: Int)
 
 }
 
 class GroupRouterSpec extends ScalaTestWithActorTestKit(GroupRouterSpec.config) with AnyWordSpecLike with LogCapturing {
   import GroupRouterSpec._
 
-  def checkGroupRouterBehavior[T](groupRouter: GroupRouter[Ping.type], settings: GroupRouterSpecSettings)(
-      resultCheck: (Seq[ActorRef[Ping.type]], Seq[ActorRef[Ping.type]]) => T): T = {
-    import scala.concurrent.duration._
-
-    import akka.actor.typed.scaladsl.AskPattern._
-    implicit val system1 =
-      createSystem(
-        Behaviors.setup[Command] { ctx =>
-          (0 until settings.node1WorkerCount).foreach { i =>
-            val worker = ctx.spawn(pingPong(ctx.self), s"ping-pong-$i")
-            ctx.system.receptionist ! Receptionist.Register(pingPongKey, worker)
-          }
-          val router = ctx.spawn(groupRouter, "group-router")
-
-          // ensure all nodes are joined and all actors are discovered.
-          ctx.system.receptionist ! Receptionist.Subscribe(
-            pingPongKey,
-            ctx.spawn(Behaviors.receiveMessage[Receptionist.Listing] {
-              case pingPongKey.Listing(update)
-                  if update.size == settings.node1WorkerCount + settings.node2WorkerCount =>
-                (0 until settings.messageCount).foreach(_ => router ! Ping)
-                Behaviors.empty
-              case _ =>
-                Behaviors.same
-            }, "waiting-actor-discovery"))
-          workerStatsBehavior(List.empty)
-        },
-        system.name)
-
-    val system2 = createSystem(Behaviors.setup[Command] { ctx =>
+  /**
+   * Starts a new pair of nodes, forms a cluster, runs a ping-pong session and hands the result to the test for verification
+   */
+  def checkGroupRouterBehavior(groupRouter: GroupRouter[PingActor.Ping], settings: GroupRouterSpecSettings): Result = {
+
+    val resultProbe = testKit.createTestProbe[Pinger.DonePinging]()
+
+    val system1 = ActorSystem(Behaviors.setup[Receptionist.Listing] {
+      ctx =>
+        (0 until settings.node1WorkerCount).foreach { i =>
+          val worker = ctx.spawn(PingActor(), s"ping-pong-$i")
+          ctx.system.receptionist ! Receptionist.Register(pingPongKey, worker)
+        }
+        val router = ctx.spawn(groupRouter, "group-router")
+        ctx.system.receptionist ! Receptionist.Subscribe(pingPongKey, ctx.self)
+
+        Behaviors.receiveMessage {
+          case pingPongKey.Listing(update) if update.size == settings.node1WorkerCount + settings.node2WorkerCount =>
+            // the requested number of workers are started and registered with the receptionist
+            // start pinging them
+            ctx.log.debug("Saw {} workers, starting pinger", update.size)
+            ctx.spawn(Pinger(router, settings.messageCount, resultProbe.ref), "pinger")
```

Comment:
If the setup is stable at this point, all workers registered and such, then I don't understand why we would need resend attempts in the Pinger? Messages are not lost under normal conditions.

Reply:
I thought I saw `Upstream failed, cause: StreamTcpException` but now I see that is a different test (RemoteAskSpec) in the same log dump. Hmmm.